### PR TITLE
ffmpeg: Build on Windows/MinGW

### PR DESCRIPTION
**Pull request by SirLynix**
Now that we have a msys2 package, we can build ffmpeg from source on Windows (which requires a shell to run configure)

### CODE REVIEW

**Comment by waruqi** [comment]
I am supporting it now. please waiting for https://github.com/xmake-io/xmake-repo/pull/3312

**Comment by SirLynix** [comment]
I don't know how to fix this
```
cl.exe -nologo -I. -I./ -D_ISOC99_SOURCE -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DWIN32_LEAN_AND_MEAN -D_USE_MATH_DEFINES -D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_WARNINGS -D_WIN32_WINNT=0x0600 -DPIC -I./compat/atomics/win32 -DHAVE_AV_CONFIG_H -DBUILDING_avdevice -I. -I./ -D_ISOC99_SOURCE -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DWIN32_LEAN_AND_MEAN -D_USE_MATH_DEFINES -D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_WARNINGS -D_WIN32_WINNT=0x0600 -DPIC -I./compat/atomics/win32 -DHAVE_AV_CONFIG_H -DBUILDING_avdevice -nologo    -W3 -wd4018 -wd4146 -wd4244 -wd4305 -wd4554 -O2   -utf-8  -showIncludes -Zs libavdevice/alldevices.c 2>&1 | awk '/including/ { sub(/^.*file: */, ""); gsub(/\\/, "/"); if (!match($0, / /)) print "libavdevice/alldevices.o:", $0 }' > libavdevice/alldevices.d
cl.exe -nologo -I. -I./ -D_ISOC99_SOURCE -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DWIN32_LEAN_AND_MEAN -D_USE_MATH_DEFINES -D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_WARNINGS -D_WIN32_WINNT=0x0600 -DPIC -I./compat/atomics/win32 -DHAVE_AV_CONFIG_H -DBUILDING_avdevice -I. -I./ -D_ISOC99_SOURCE -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DWIN32_LEAN_AND_MEAN -D_USE_MATH_DEFINES -D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_WARNINGS -D_WIN32_WINNT=0x0600 -DPIC -I./compat/atomics/win32 -DHAVE_AV_CONFIG_H -DBUILDING_avdevice -nologo    -W3 -wd4018 -wd4146 -wd4244 -wd4305 -wd4554 -O2   -utf-8  -showIncludes -Zs libavdevice/avdevice.c 2>&1 | awk '/including/ { sub(/^.*file: */, ""); gsub(/\\/, "/"); if (!match($0, / /)) print "libavdevice/avdevice.o:", $0 }' > libavdevice/avdevice.d
cl.exe -nologo -I. -I./ -D_ISOC99_SOURCE -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DWIN32_LEAN_AND_MEAN -D_USE_MATH_DEFINES -D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_WARNINGS -D_WIN32_WINNT=0x0600 -DPIC -I./compat/atomics/win32 -DHAVE_AV_CONFIG_H -DBUILDING_avdevice -I. -I./ -D_ISOC99_SOURCE -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DWIN32_LEAN_AND_MEAN -D_USE_MATH_DEFINES -D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_WARNINGS -D_WIN32_WINNT=0x0600 -DPIC -I./compat/atomics/win32 -DHAVE_AV_CONFIG_H -DBUILDING_avdevice -nologo    -W3 -wd4018 -wd4146 -wd4244 -wd4305 -wd4554 -O2   -utf-8  -showIncludes -Zs libavdevice/dshow.c 2>&1 | awk '/including/ { sub(/^.*file: */, ""); gsub(/\\/, "/"); if (!match($0, / /)) print "libavdevice/dshow.o:", $0 }' > libavdevice/dshow.d
cl.exe -nologo -I. -I./ -D_ISOC99_SOURCE -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DWIN32_LEAN_AND_MEAN -D_USE_MATH_DEFINES -D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_WARNINGS -D_WIN32_WINNT=0x0600 -DPIC -I./compat/atomics/win32 -DHAVE_AV_CONFIG_H -DBUILDING_avdevice -I. -I./ -D_ISOC99_SOURCE -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DWIN32_LEAN_AND_MEAN -D_USE_MATH_DEFINES -D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_WARNINGS -D_WIN32_WINNT=0x0600 -DPIC -I./compat/atomics/win32 -DHAVE_AV_CONFIG_H -DBUILDING_avdevice -nologo    -W3 -wd4018 -wd4146 -wd4244 -wd4305 -wd4554 -O2   -utf-8  -showIncludes -Zs libavdevice/dshow_common.c 2>&1 | awk '/including/ { sub(/^.*file: */, ""); gsub(/\\/, "/"); if (!match($0, / /)) print "libavdevice/dshow_common.o:", $0 }' > libavdevice/dshow_common.d
cl.exe -nologo -I. -I./ -D_ISOC99_SOURCE -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DWIN32_LEAN_AND_MEAN -D_USE_MATH_DEFINES -D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_WARNINGS -D_WIN32_WINNT=0x0600 -DPIC -I./compat/atomics/win32 -DHAVE_AV_CONFIG_H -DBUILDING_avdevice -I. -I./ -D_ISOC99_SOURCE -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DWIN32_LEAN_AND_MEAN -D_USE_MATH_DEFINES -D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_WARNINGS -D_WIN32_WINNT=0x0600 -DPIC -I./compat/atomics/win32 -DHAVE_AV_CONFIG_H -DBUILDING_avdevice -nologo    -W3 -wd4018 -wd4146 -wd4244 -wd4305 -wd4554 -O2   -utf-8  -showIncludes -Zs libavdevice/dshow_crossbar.c 2>&1 | awk '/including/ { sub(/^.*file: */, ""); gsub(/\\/, "/"); if (!match($0, / /)) print "libavdevice/dshow_crossbar.o:", $0 }' > libavdevice/dshow_crossbar.d
cl.exe -nologo -I. -I./ -D_ISOC99_SOURCE -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DWIN32_LEAN_AND_MEAN -D_USE_MATH_DEFINES -D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_WARNINGS -D_WIN32_WINNT=0x0600 -DPIC -I./compat/atomics/win32 -DHAVE_AV_CONFIG_H -DBUILDING_avdevice -I. -I./ -D_ISOC99_SOURCE -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DWIN32_LEAN_AND_MEAN -D_USE_MATH_DEFINES -D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_WARNINGS -D_WIN32_WINNT=0x0600 -DPIC -I./compat/atomics/win32 -DHAVE_AV_CONFIG_H -DBUILDING_avdevice -nologo    -W3 -wd4018 -wd4146 -wd4244 -wd4305 -wd4554 -O2   -utf-8  -showIncludes -Zs libavdevice/dshow_enummediatypes.c 2>&1 | awk '/including/ { sub(/^.*file: */, ""); gsub(/\\/, "/"); if (!match($0, / /)) print "libavdevice/dshow_enummediatypes.o:", $0 }' > libavdevice/dshow_enummediatypes.d
awk: cmd. line:1: /including/ { sub(/^.*file: */, ""); gsub(/\/, "/"); if (!match($0, / /)) print "libavdevice/alldevices.o:", $0 }
awk: cmd. line:1:                                                                                                        ^ syntax error
awk: cmd. line:1: /including/ { sub(/^.*file: */, ""); gsub(/\/, "/"); if (!match($0, / /)) print "libavdevice/alldevices.o:", $0 }
awk: cmd. line:1:                                                                                                           ^ unterminated string
awk: cmd. line:1: /including/ { sub(/^.*file: */, ""); gsub(/\/, "/"); if (!match($0, / /)) print "libavdevice/dshow.o:", $0 }
awk: cmd. line:1:                                                                                                   ^ syntax error
awk: cmd. line:1: /including/ { sub(/^.*file: */, ""); gsub(/\/, "/"); if (!match($0, / /)) print "libavdevice/dshow.o:", $0 }
awk: cmd. line:1:                                                                                                      ^ unterminated string
awk: cmd. line:1: /including/ { sub(/^.*file: */, ""); gsub(/\/, "/"); if (!match($0, / /)) print "libavdevice/dshow_crossbar.o:", $0 }
awk: cmd. line:1:                                                                                                            ^ syntax error
awk: cmd. line:1: /including/ { sub(/^.*file: */, ""); gsub(/\/, "/"); if (!match($0, / /)) print "libavdevice/dshow_crossbar.o:", $0 }
awk: cmd. line:1:                                                                                                               ^ unterminated string
awk: cmd. line:1: /including/ { sub(/^.*file: */, ""); gsub(/\/, "/"); if (!match($0, / /)) print "libavdevice/dshow_common.o:", $0 }
awk: cmd. line:1:                                                                                                          ^ syntax error
awk: cmd. line:1: /including/ { sub(/^.*file: */, ""); gsub(/\/, "/"); if (!match($0, / /)) print "libavdevice/dshow_common.o:", $0 }
awk: cmd. line:1:                                                                                                             ^ unterminated string
awk: cmd. line:1: /including/ { sub(/^.*file: */, ""); gsub(/\/, "/"); if (!match($0, / /)) print "libavdevice/dshow_enummediatypes.o:", $0 }
awk: cmd. line:1:                                                                                                                  ^ syntax error
awk: cmd. line:1: /including/ { sub(/^.*file: */, ""); gsub(/\/, "/"); if (!match($0, / /)) print "libavdevice/dshow_enummediatypes.o:", $0 }
awk: cmd. line:1:                                                                                                                     ^ unterminated string
awk: cmd. line:1: /including/ { sub(/^.*file: */, ""); gsub(/\/, "/"); if (!match($0, / /)) print "libavdevice/avdevice.o:", $0 }
awk: cmd. line:1:                                                                                                      ^ syntax error
awk: cmd. line:1: /including/ { sub(/^.*file: */, ""); gsub(/\/, "/"); if (!match($0, / /)) print "libavdevice/avdevice.o:", $0 }
awk: cmd. line:1:                                                                                                         ^ unterminated string
```

I'm running make using bash.exe but it seems make tries to run commands with cmd.exe, is there a way to tell it which shell to use? I already overrode SHELL env

**Comment by waruqi** [comment]
I don't know too.

**Comment by SirLynix** [comment]
I fixed it, it was because of msvc:runenvs which included PATH, giving it priority over msys2 binaries

**Comment by SirLynix** [comment]
Almost there.

**Comment by waruqi** [comment]
> ld: unknown option: --nxcompat

we can apply patch to remove these flags if they are not supported.

File: packages/f/ffmpeg/xmake.lua [hunk 1]
```diff
@@ -3,63 +3,55 @@ package("ffmpeg")
...
-        add_configs("bzlib",            {description = "Enable bzlib compression library.", default = false, type = "boolean"})
-        add_configs("libx264",          {description = "Enable libx264 decoder.", default = false, type = "boolean"})
-        add_configs("libx265",          {description = "Enable libx265 decoder.", default = false, type = "boolean"})
-        add_configs("iconv",            {description = "Enable libiconv library.", default = false, type = "boolean"})
-        add_configs("vaapi",            {description = "Enable vaapi library.", default = false, type = "boolean"})
-        add_configs("vdpau",            {description = "Enable vdpau library.", default = false, type = "boolean"})
-        add_configs("hardcoded-tables", {description = "Enable hardcoded tables.", default = true, type = "boolean"})
+    if is_plat("windows") and is_arch("arm", "arm64") then
+        add_resources(">=1.0.0", "gas-preprocessor", "https://raw.githubusercontent.com/FFmpeg/gas-preprocessor/master/gas-preprocessor.pl", "17a14de55348547d5de5c2448dee0ab6d193d138a1326b6a5235617c3f2c2f82")
```
Comment (waruqi) [inline packages/f/ffmpeg/xmake.lua:11]: we need not download it, xmake has this script.

```
path.join(os.programdir(), "scripts", "gas-preprocessor.pl")
```

File: packages/f/ffmpeg/xmake.lua [hunk 2]
```diff
@@ -153,7 +135,86 @@ package("ffmpeg")
...
+            envs.SHELL = "sh"
+
+            if package:is_arch("arm", "arm64") then
+                os.mkdir("gas-preprocessor")
+                os.vmv(package:resourcefile("gas-preprocessor"), path.join("gas-preprocessor", "gas-preprocessor.pl"))
+                envs.PATH = path.join(os.curdir(), "gas-preprocessor") .. path.envsep() .. envs.PATH
+            end
+
+            table.insert(configs, "--prefix=" .. package:installdir())
+            os.vrunv("sh", table.join("./configure", configs), {envs = envs})
```
Comment (waruqi) [inline packages/f/ffmpeg/xmake.lua:200]: `{shell = true}`

File: packages/f/ffmpeg/xmake.lua [hunk 3]
```diff
@@ -221,18 +277,29 @@ package("ffmpeg")
...
             os.vrunv("make", argv)
             os.vrun("make install")
         else
-            if package:is_cross() then
-                table.insert(configs, "--enable-cross-compile")
-                table.insert(configs, "--arch=" .. package:targetarch())
-                configs.host = "" -- prevents xmake to add a --host=xx parameter
+            if package:is_plat("mingw") and package:is_cross() then
+                -- looks like MinGW on macOS doesn't support -Wl,--pic-executable
+                io.replace("configure", "add_ldexeflags -Wl,--pic-executable,-e,_mainCRTStartup", "add_ldexeflags -Wl,-e,_mainCRTStartup", {plain=true})
```
Comment (waruqi) [inline packages/f/ffmpeg/xmake.lua:288]: `plain = true`, add spaces

File: packages/f/ffmpeg/xmake.lua [hunk 4]
```diff
@@ -191,13 +187,12 @@ package("ffmpeg")
             envs.SHELL = "sh"
 
             if package:is_arch("arm", "arm64") then
-                os.mkdir("gas-preprocessor")
-                os.vmv(package:resourcefile("gas-preprocessor"), path.join("gas-preprocessor", "gas-preprocessor.pl"))
-                envs.PATH = path.join(os.curdir(), "gas-preprocessor") .. path.envsep() .. envs.PATH
+                -- for gas-preprocessor.pl
+                envs.PATH = path.join(os.programdir(), "scripts") .. path.envsep() .. envs.PATH
```
Comment (waruqi) [inline packages/f/ffmpeg/xmake.lua:191]: use path.joinenv

Comment (waruqi) [reply packages/f/ffmpeg/xmake.lua:191]: and we should use table. os.execv will call path.joinenv

File: packages/f/ffmpeg/xmake.lua [hunk 5]
```diff
@@ -153,7 +131,85 @@ package("ffmpeg")
...
+        if package:is_cross() then
+            table.insert(configs, "--enable-cross-compile")
+            table.insert(configs, "--arch=" .. package:targetarch())
+        end
+
+        -- build
+        if package:is_plat("windows") then
+            import("core.base.option")
+            import("core.tool.toolchain")
+            local msvc = toolchain.load("msvc", {plat = package:plat(), arch = package:arch()})
```
Comment (waruqi) [inline packages/f/ffmpeg/xmake.lua:184]: and here, package:toolchain

**Comment by SirLynix** [comment]
```
ld: warning: could not create compact unwind for _kbd_window_init: stack subq instruction is too different from dwarf stack size
ld: in libswresample/libswresample.a(audio_convert.o), archive member 'audio_convert.o' with length 40552 is not mach-o or llvm bitcode file 'libswresample/libswresample.a' for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

I have no idea how to fix this

**Comment by waruqi** [comment]
> ```
> ld: warning: could not create compact unwind for _kbd_window_init: stack subq instruction is too different from dwarf stack size
> ld: in libswresample/libswresample.a(audio_convert.o), archive member 'audio_convert.o' with length 40552 is not mach-o or llvm bitcode file 'libswresample/libswresample.a' for architecture x86_64
> clang: error: linker command failed with exit code 1 (use -v to see invocation)
> ```
> 
> I have no idea how to fix this

you can disable mingw@macosx

File: packages/f/ffmpeg/scripts/gas-preprocessor.pl [hunk 6]
```diff
@@ -0,0 +1,1243 @@
+#!/usr/bin/env perl
+# by David Conrad
```
Comment (waruqi) [inline packages/f/ffmpeg/scripts/gas-preprocessor.pl:2]: it's too large, please use xmake's script.

Comment (SirLynix) [reply packages/f/ffmpeg/scripts/gas-preprocessor.pl:2]: But it's only available on xmake dev, isn't that an issue?

Comment (waruqi) [reply packages/f/ffmpeg/scripts/gas-preprocessor.pl:2]: Before xmake is released, you can download it as a resource first

Comment (SirLynix) [reply packages/f/ffmpeg/scripts/gas-preprocessor.pl:2]: I'm not on my computer for now, I will change it asap

**Pull request merged**